apollo-router-core/src/request.rs

```diff
@@ -44,34 +45,81 @@ where
     <Option<T>>::deserialize(deserializer).map(|x| x.unwrap_or_default())
 }
 
+#[derive(Deserialize)]
+struct RequestMeta {
+    query: String,
+    operation_name: Option<String>,
+}
```

Comment:
:grey_question: I'm not sure why you used 2 different ways to extract the data from the request's bytes here.

At first you extract "variables" and "extensions" from the value manually. Then the rest of the value is deserialized to RequestMeta.

I would expect that one is using serde_json and the other is using serde_json_bytes but since Value is imported from json_ext which uses serde_json_bytes, it looks like everything is identical.

Why not just either extract all the fields manually, either through this struct? It would make the code more straightforward to read.

Reply:
so this is something I tried in https://github.com/apollographql/router/pull/284/commits/e85bf735cb1ec37176994faa2a54c6e183b38598 and while it looks like it does what we want, we'd lose all of the performance benefits, because it will go through all of the data in `Value`, pass it through the deserializer, and convert them, but there's no way there to handle the `ByteString` type. So if there's a string that is stored as a `ByteString`, the deserializer will get it as `&str`, then when reconstructing the `Value` allocate a new `ByteString`.

That's why here I have to separate the parts of the struct that I can deserialize directly with sede, and the extensions and variables that should be just moved and not modified. For the request I could just deserialize everything manually, but for `Response` and `FetchError`, I'd rather avoid writing manually the deserialization for locations

Reply:
> I'd rather avoid writing manually the deserialization for locations

I made a macro specifically for that that could be useful: https://crates.io/crates/serde_json_traversal

> So if there's a string that is stored as a ByteString, the deserializer will get it as &str, then when reconstructing the Value allocate a new ByteString.

Oh ok I think I see... well at least this should be explained in comment: there is a bit of important logic here.

If you don't mind I can take another look (having a second perspective) on that particular issue and see if I find another way.